### PR TITLE
blender: 2.93.5 -> 3.0.0, keep old as blender_2_93, refactor, support wayland

### PR DESCRIPTION
**Pull request by kanashimia**
###### Motivation for this change
https://www.blender.org/download/releases/3-0/

Blender 3.0 removed support for OpenCL rendering, so we should probably keep old version for now.  

OptiX support had to be removed because publicly available version is too old.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix
Comment (veprbl) [inline pkgs/top-level/all-packages.nix:24223]:
```suggestion
  blender_2_93 = callPackage  ../applications/misc/blender/2.93.nix {
```

or

```suggestion
  blender_2 = callPackage  ../applications/misc/blender/2.nix {
```

**Review by veprbl — COMMENTED** [review]
Is there a reason to keep blender 2.x? If 3.0 builds on darwin, we can just remove 2.x

File: pkgs/top-level/all-packages.nix
Comment (kanashimia) [reply pkgs/top-level/all-packages.nix:24223]: > Is there a reason to keep blender 2.x? If 3.0 builds on darwin, we can just remove 2.x

Blender 3.0 removed support for OpenCL rendering, which i bet many of the amd gpu users would miss.

File: pkgs/applications/misc/blender/default.nix
Comment (veprbl) [inline pkgs/applications/misc/blender/default.nix:103]:
```suggestion
  '';
```

File: pkgs/applications/misc/blender/2.93.nix
Comment (veprbl) [inline pkgs/applications/misc/blender/2.93.nix:11]:
```suggestion
, openxr-loader, openvdb, openal
```

**Review by veprbl — COMMENTED** [review]
https://code.blender.org/2021/11/next-level-support-for-amd-gpus/#:~:text=Blender%203.0%20was%20announced%20with,for%20technical%20and%20performance%20reasons.

**Comment by kanashimia** [comment]
It would be nice to know if AMD HIP support actually works for peopele with RDNA 2 cards.

File: pkgs/applications/misc/blender/2.93.nix
Comment (kanashimia) [reply pkgs/applications/misc/blender/2.93.nix:11]: Note that this should be in `default.nix` instead of `2.93.nix` i assume.

Comment (veprbl) [reply pkgs/applications/misc/blender/2.93.nix:11]: Right

File: pkgs/applications/misc/blender/default.nix
Comment (veprbl) [inline pkgs/applications/misc/blender/default.nix:47]: The benefit of using `fetchurl` is that it allows for tarballs to be mirrored on tarballs.nixos.org

**Review by veprbl — APPROVED** [review]

**Comment by minijackson** [comment]
is it possible to enable wayland support by default? That way, wayland users can install a wayland blender without having to compile it. Alternatively a `blender-wayland` package should be possible?

**Comment by Philipp-M** [comment]
Just tested this with CUDA enabled, with the overlay:
```nix
nixpkgs-blender3.pkgs.blender.override {
  cudaSupport = true;
  # cudatoolkit = nixpkgs-unstable.pkgs.cudatoolkit_11_4;
}
```

I had to replace `cudatoolkit` with the version above to be able to build it, as otherwise this error occurs (my gpu might be to new for the default cudatoolkit?):

```
...
make[2]: *** [intern/cycles/kernel/CMakeFiles/cycles_kernel_cuda.dir/build.make:2228: intern/cycles/kernel/kernel_sm_75.cubin] Error 139
...
make[1]: *** [CMakeFiles/Makefile2:3483: intern/cycles/kernel/CMakeFiles/cycles_kernel_cuda.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
```
(I can upload the full log, if it's relevant)

But although it compiles and runs with CUDA support, the Optix tab doesn't show up in the preferences under `System -> Cycles Render Devices`

blender via flatpak/flathub shows Optix tab for reference.

It might be useful to probably add a `blenderWithCuda` package with a newer cudatoolkit? And/or update the default cudatoolkit alltogether by that matter?

**Comment by jansol** [comment]
+1 for building with wayland enabled by default. Unless that does something "funny", like disable X11 as the build system of Renderdoc does, in which case a separate blender-wayland package would be much appreciated.

As for HIP, even if it worked on the Blender side, it requires ROCm which still doesn't officially advertise support for anything past the Radeon VII, and I have yet to manage to run anything more complex than `clinfo` on it without crashing. Can't wait for Blender to finally get that Vulkan renderer.

**Comment by kanashimia** [comment]
> +1 for building with wayland enabled by default. Unless that does something "funny", like disable X11 as the build system of Renderdoc does, in which case a separate blender-wayland package would be much appreciated.

Strongly against this unless it becames default upstream, it is disabled for a reason.
I havent used it for a much, but already found some bugs that render it unusable, like it doesn't bind tablet button handlers, and so you can't rotate viewport using pen.

File: pkgs/applications/misc/blender/default.nix
Comment (kanashimia) [reply pkgs/applications/misc/blender/default.nix:47]: But fetchzip is a wrapper around fetchurl, how does that work?

Comment (veprbl) [reply pkgs/applications/misc/blender/default.nix:47]: https://github.com/NixOS/nixpkgs/blob/2f218ac01d9d1c1ef8cca1cd0eb13d4971209b89/maintainers/scripts/find-tarballs.nix#L19-L23

`drv.outputHashMode == "flat"` is required for some reason

**Comment by kanashimia** [comment]
#### Bad news:
OptiX support has to be removed (for now)

blender 3.0 requires OptiX 7.3 or later, but to download it one has to register on the nvidia developer portal, and we won't be able to redistribute it.
Archlinux removed it: https://github.com/archlinux/svntogit-community/blob/packages/blender/trunk/PKGBUILD#L57

Perhaps we could package blender-bin in the future with it added.

#### Good news:
You can run blender without cudaSupport and have cuda support.
That option is extremely misleading, it took long to figure out for me.
What it does is makes blender compile all the cuda kernels at the compile time.

By default blender compiles itself with dynload cuda support, it will try to dlopen the libcuda.so on the runtime, then will try to compile cuda kernel using nvcc first time you try to render with cuda (takes some time).

I was able to run blender with cuda like this:
```
NIXPKGS_ALLOW_UNFREE=1 nix shell n#cudatoolkit_11 --impure
export CYCLES_CUDA_EXTRA_CFLAGS="-I$(NIXPKGS_ALLOW_UNFREE=1 nix eval --raw n#cudatoolkit_11 --impure)/include"
export LD_LIBRARY_PATH=/run/opengl-driver/lib 
blender
```

So only thing have left is to figure out what to do so that blender won't be rebuilt.
I believe it is possible to spilt blender into two derivations, wrapped and unwrapped, then blender itself wont be rebuilt when you want to use cuda.

**Comment by kanashimia** [comment]
> It might be useful to probably add a `blenderWithCuda` package with a newer cudatoolkit? And/or update the default cudatoolkit alltogether by that matter?
Indeed, blender 3.0 requires cudatoolkit 11.1 or later as it says on the wiki.

File: pkgs/applications/misc/blender/default.nix
Comment (pennae) [inline pkgs/applications/misc/blender/default.nix:44]: that's not the right hash any more

Comment (FedX-sudo) [reply pkgs/applications/misc/blender/default.nix:44]: Ah, glad I am not the only one.

Comment (kanashimia) [reply pkgs/applications/misc/blender/default.nix]: Sorry, my fault, i've messed up when rebasing, ugh :1st_place_medal:

**Comment by kanashimia** [comment]
Okay, wrapper seems to work, i'm not sure at all how to be with darwin, i can't (not too bothered to) test it, was it even working before? :moyai:  

Also, ofborg figured out that this merge request has a clean-up? damn boy is smart

**Review by pennae — APPROVED** [review]
can't test with wayland, but seems to work fine with X.

File: pkgs/applications/misc/blender/default.nix
Comment (veprbl) [inline pkgs/applications/misc/blender/default.nix:143]: Looks like a misuse of `symlinkJoin`

File: pkgs/applications/misc/blender/default.nix
Comment (veprbl) [inline pkgs/applications/misc/blender/default.nix:146]:
```
runCommandNoCC {
# ...

  unwrapped = blender-unwrapped;
  buildCommand = ''
    wrapProgram $unwrapped/foo/bar/blender
  '';
}
```
Allows to do things like
```nix
blender.overrideAttrs (_: { unwrapped = blender.unwrapped.overrideAttrs (_: { /* src = ...; */ }); })
```
As well as simply accessing the unwrapped derivation

See prior PR #97388

**Comment by veprbl** [comment]
Should we get this in without the wrapper?

**Comment by pennae** [comment]
would be appreciated! there is the blender_bin flake after all, though no idea how usable that is from a non-flakes system.

**Comment by kanashimia** [comment]
Ehhh, rebasing this on top of blender-with-wrapper is hard :angry:

**Comment by MatthewCroughan** [comment]
@kanashimia 

> blender 3.0 requires OptiX 7.3 or later

This is not the case. I tried to compile with 7.4 but get the following errors. Using 7.3 works fine. It needs to be exactly 7.3, even according to the docs here https://wiki.blender.org/wiki/Building_Blender/CUDA

```
blender> /build/blender-3.0.0/intern/cycles/device/optix/device_impl.cpp: In member function 'virtual bool ccl::OptiXDevice::load_kernels(ccl::uint)':
blender> /build/blender-3.0.0/intern/cycles/device/optix/device_impl.cpp:211:33: error: 'OPTIX_COMPILE_DEBUG_LEVEL_LINEINFO' was not declared in this scope; did you mean 'OPTIX_COMPILE_DEBUG_LEVEL_MINIMAL'?
blender>   211 |     module_options.debugLevel = OPTIX_COMPILE_DEBUG_LEVEL_LINEINFO;
blender>       |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
blender>       |                                 OPTIX_COMPILE_DEBUG_LEVEL_MINIMAL
blender> /build/blender-3.0.0/intern/cycles/device/optix/device_impl.cpp:414:31: error: 'OPTIX_COMPILE_DEBUG_LEVEL_LINEINFO' was not declared in this scope; did you mean 'OPTIX_COMPILE_DEBUG_LEVEL_MINIMAL'?
blender>   414 |     link_options.debugLevel = OPTIX_COMPILE_DEBUG_LEVEL_LINEINFO;
blender>       |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
blender>       |                               OPTIX_COMPILE_DEBUG_LEVEL_MINIMAL
blender> [ 71%] Building CXX object source/blender/compositor/CMakeFiles/bf_compositor.dir/nodes/COM_DisplaceNode.cc.o
blender> [ 71%] Building C object source/blender/makesrna/intern/CMakeFiles/makesrna.dir/rna_material_api.c.o
blender> /build/blender-3.0.0/intern/cycles/device/optix/device_impl.cpp: In member function 'virtual void ccl::OptiXDevice::build_bvh(ccl::BVH*, ccl::Progress&, bool)':
blender> /build/blender-3.0.0/intern/cycles/device/optix/device_impl.cpp:1498:27: error: 'OPTIX_INSTANCE_FLAG_DISABLE_TRANSFORM' was not declared in this scope; did you mean 'OPTIX_INSTANCE_FLAG_DISABLE_ANYHIT'?
blender>  1498 |         instance.flags |= OPTIX_INSTANCE_FLAG_DISABLE_TRANSFORM;
blender>       |                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
blender>       |                           OPTIX_INSTANCE_FLAG_DISABLE_ANYHIT
blender> [ 71%] Building CXX object source/blender/compositor/CMakeFiles/bf_compositor.dir/nodes/COM_MapUVNode.cc.o
blender> [ 71%] Building C object source/blender/makesrna/intern/CMakeFiles/makesrna.dir/rna_mesh_api.c.o
blender> /build/blender-3.0.0/intern/cycles/device/optix/device_impl.cpp:1515:29: error: 'OPTIX_INSTANCE_FLAG_DISABLE_TRANSFORM' was not declared in this scope; did you mean 'OPTIX_INSTANCE_FLAG_DISABLE_ANYHIT'?
blender>  1515 |           instance.flags |= OPTIX_INSTANCE_FLAG_DISABLE_TRANSFORM;
```

**Comment by InternetUnexplorer** [comment]
This might be old news by now, but the OptiX 7.3 headers are available [here](https://developer.download.nvidia.com/redist/optix/v7.3/OptiX-7.3.0-Include.zip) now. :tada:

Also, is there any way I can help out with this? Blender 3.1 is going to be out soon and it would be really nice to have a 3.x version in Nixpkgs! :)

**Comment by tobiasBora** [comment]
For people wanting to use blender 3.0 without waiting for this PR to be merged, you can also simply use the bin provided by the blender project and run (after uncompressing the `.tar.gz` file) it into `steam-run`, it runs fine for what I have tried:
```
$ steam-run ./blender
```
Of course, I can't wait to see this PR officially merged.

**Comment by matthew-piziak** [comment]
@tobiasBora I've been using the following expression:

```
blender = pkgs.blender.overrideAttrs (orig: rec {
    version = "3.0.0";
    src = pkgs.fetchurl {
      url = "https://download.blender.org/source/${orig.pname}-${version}.tar.xz";
      sha256 = "1jzirg60c2lhln78a7phbsk2ssvcdqxqb3awp895m0pqrlmz7w2h";
    };
    buildInputs = orig.buildInputs ++ (with pkgs; [ zstd ]);
  });
```

**Comment by tobiasBora** [comment]
@mattew-piziak Really? What's the point of this PR if it's possible to compile blender 3 by just changing the url of the current derivation?

Btw, I thought that blender had a `blender-bin` derivation using the binary files provided by the Blender foundation, but I can't find it using `nix search`, am I missing something?

**Comment by kirillrdy** [comment]
@tobiasBora if you use flakes

`nix search blender-bin blender`

**Comment by matthew-piziak** [comment]
@tobiasBora No idea, that's beyond my paygrade. I'm just monitoring this PR so I know when I can switch to the upstream derivation. I did need to add `zstd` to make it build, but I don't know what else might be necessary for a proper derivation.

**Comment by pennae** [comment]
[blender 3.1 has been released](https://www.blender.org/download/releases/3-1/)

**Comment by mkaito** [comment]
> [blender 3.1 has been released](https://www.blender.org/download/releases/3-1/)

And another sad day for AMD GPU owners

**Comment by matthew-piziak** [comment]
@mkaito Why, what happened in 3.1?

**Comment by mkaito** [comment]
> @mkaito Why, what happened in 3.1?

Still no HIP support in Cycles on Linux at all (soon tm, maybe), and support for HIP on cards older than RDNA1 is "being researched, but very low priority", according to an unofficial AMD insider source about 2 days ago.

This was all supposed to land in 3.1. Now maybe HIP on Linux will land in 3.2. If you have a GCN5 or older card, maybe never.

**Comment by jansol** [comment]
> What's the point of this PR if it's possible to compile blender 3 by just changing the url of the current derivation?

From the PR title: "... keep old as blender_2_93, refactor, support wayland"

File: pkgs/applications/misc/blender/default.nix
Comment (jansol) [inline pkgs/applications/misc/blender/default.nix:29]: Is wayland support mutually exclusive with X11 support? if not, I'd rather default this to `stdenv.isLinux`

Comment (minijackson) [reply pkgs/applications/misc/blender/default.nix:29]: From @kanashimia's comment (https://github.com/NixOS/nixpkgs/pull/148550#issuecomment-986272008), wayland support is quite experimental and disabled by default upstream.

**Comment by tobiasBora** [comment]
Just to let you know, blender got updated to 3.1 in https://github.com/NixOS/nixpkgs/pull/164987 I'm not sure what's the status of this PR now.

**Comment by necrophcodr** [comment]
> > @mkaito Why, what happened in 3.1?
> 
> Still no HIP support in Cycles on Linux at all (soon tm, maybe), and support for HIP on cards older than RDNA1 is "being researched, but very low priority", according to an unofficial AMD insider source about 2 days ago.
> 
> This was all supposed to land in 3.1. Now maybe HIP on Linux will land in 3.2. If you have a GCN5 or older card, maybe never.

 Blender 3.2 recently launched, which enables this Linux rendering on RDNA and RDNA 2 cards. It doesn't seem to be the case that anything older is being implemented yet.

**Comment by InternetUnexplorer** [comment]
ROCm support will require at least replacing some [hardcoded paths](https://github.com/blender/blender/blob/dfb8c90324c018068f12ca41543c3c311dff6d21/extern/hipew/src/hipew.c#L256), possibly more; I wasn't able to get it working when I tried it a month or two ago (but ROCm has gone from 5.1.0 → 5.1.3 since then, so maybe it will work now). I think a PR for Blender 3.1 would be a better place to discuss it, though.

**Comment by InternetUnexplorer** [comment]
Oh, sorry, I meant 3.2 😅

**Comment by NickCao** [comment]
superseded by later updates.

**Pull request closed without merging**
